R/tags.R

```diff
@@ -1182,14 +1204,37 @@ withTags <- function(code) {
 
 # Make sure any objects in the tree that can be converted to tags, have been
 tagify <- function(x) {
-  rewriteTags(x, function(uiObj) {
-    if (isResolvedTag(uiObj) || isTagList(uiObj) || is.character(uiObj))
-      return(uiObj)
-    else
-      tagify(as.tags(uiObj))
+
+  # Schedule .postRenderHook(s) to run _after_ converting to tags
+  postHooks <- list()
+  on.exit({
+    for (hook in postHooks) tryWarn(hook())
+  })
+
+  rewriteTags(x, function(ui) {
+    if (is.character(ui)) return(ui)
+
+    pre <- attr(ui, "renderHooks")
+    for (hook in pre) {
+      tryWarn(ui <- as.tags(hook(ui)))
+    }
+    attr(ui, "renderHooks") <- NULL
+
+    post <- attr(ui, "postRenderHooks")
+    if (length(post)) {
+      postHooks <<- c(postHooks, post)
+      attr(ui, "postRenderHooks") <- NULL
+    }
+
+    if (isTag(ui) || isTagList(ui)) ui else tagify(as.tags(ui))
+
   }, FALSE)
 }
 
+tryWarn <- function(expr) {
+  tryCatch(expr, error = function(e) warning(conditionMessage(e)))
+}
```

Comment:
I think we'll want a better warning message here. For example:

```
> tryWarn(x <- stop("xyz"))
Warning message:
In value[[3L]](cond) : xyz
```


Maybe something like this:

```R
tryWarn <- function(expr) {
  result <- NULL
  error_message <- NULL
  result <- tryCatch({ expr }, error = function(e) error_message <<- conditionMessage(e))
  
  if (is.null(error_message)) {
    return(result)
  } else {
    warning(error_message)
    invisible(NULL)
  }
}

tryWarn(x <- stop("xyz"))
#> Warning message:
#> In tryWarn(x <- stop("xyz")) : xyz
```

However, one kind of strange thing about this is that in the case of an error, it (invisibly) returns `NULL`. I'm not sure what a better return value would be, though.